traffic_monitor/health/cache.go

```diff
@@ -144,177 +191,153 @@ func EvalCache(result cache.ResultInfo, resultStats *threadsafe.ResultStatValHis
 	}
+	return func(cache.AvailableTuple, tc.TrafficServer) bool { return false }
+}
+
+// CalcAvailability calculates the availability of each cache in results.
+// statResultHistory may be nil, in which case stats won't be used to calculate
+// availability.
+func CalcAvailability(
+	results []cache.Result,
+	pollerName string,
+	statResultHistory *threadsafe.ResultStatHistory,
+	mc tc.TrafficMonitorConfigMap,
+	toData todata.TOData,
+	localCacheStatusThreadsafe threadsafe.CacheAvailableStatus,
+	localStates peer.CRStatesThreadsafe,
+	events ThreadsafeEvents,
+	protocol config.PollingProtocol,
+) {
+	localCacheStatuses := localCacheStatusThreadsafe.Get().Copy()
+	var statResultsVal *threadsafe.CacheStatHistory
+	processAvailableTuple := getProcessAvailableTuple(protocol)
 
 	for _, result := range results {
 		if statResultHistory != nil {
-			t := statResultHistory.LoadOrStore(tc.CacheName(result.ID))
+			t := statResultHistory.LoadOrStore(result.ID)
 			statResultsVal = &t
 		}
 		serverInfo, ok := mc.TrafficServer[result.ID]
 		if !ok {
 			log.Errorf("Cache %v missing from from Traffic Ops Monitor Config - treating as OFFLINE\n", result.ID)
 		}
 
-		if _, ok := localCacheStatuses[tc.CacheName(result.ID)]; !ok {
-			localCacheStatuses[tc.CacheName(result.ID)] = make(map[string]cache.AvailableStatus)
+		availStatus := cache.AvailableStatus{
+			LastCheckedIPv4:    result.UsingIPv4,
+			ProcessedAvailable: true,
+			Poller:             pollerName,
+			Status:             serverInfo.ServerStatus,
 		}
 
+		lastStatus, ok := localCacheStatuses[result.ID]
+		if ok {
+			availStatus.Available = lastStatus.Available
+		}
+
+		reasons := []string{}
 		resultInfo := cache.ToInfo(result)
-		for interfaceName, _ := range result.Interfaces() {
-			if statResultsVal != nil {
-				t := (*statResultsVal)[interfaceName]
-				statResults = &t
+		for _, inf := range serverInfo.Interfaces {
+			if !inf.Monitor {
+				continue
 			}
-			isAvailable, usingIPv4, whyAvailable, unavailableStat := EvalCache(resultInfo, statResults, &mc)
-
-			// if the cache is now Available, and was previously unavailable due to a threshold, make sure this poller contains the stat which exceeded the threshold.
-			previousStatus, hasPreviousStatus := localCacheStatuses[tc.CacheName(result.ID)][interfaceName]
-			availableTuple := cache.AvailableTuple{}
-
-			if hasPreviousStatus {
-				availableTuple = previousStatus.Available
-				availableTuple.SetAvailability(usingIPv4, isAvailable)
-
-				if processAvailableTuple(availableTuple, serverInfo) {
-					if !processAvailableTuple(previousStatus.Available, serverInfo) && previousStatus.UnavailableStat != "" {
-						if !result.HasStat(previousStatus.UnavailableStat) {
-							return
-						}
-					}
-				}
+
+			available, why := EvalInterface(resultInfo.InterfaceVitals, inf)
+			if result.UsingIPv4 {
+				availStatus.Available.IPv4 = availStatus.Available.IPv4 && available
```

Comment:
with the last change and this line, if the previous status is unavailable then it can never get to be available. ill dig deeper today too to come up with another solution but wanted to throw it out there

Reply:
Does this even need the `availStatus.Available.IPvX &&`? 

Also if I'm set to using IPv4, and in the next cycle I change to using IPv6, Available.IPv4 will remain `true` even though it shouldn't be.

Reply:
well based on what Matt said, I think it should remain true:

>  _"the way this is written, IPv4 and IPv6 will never both be labeled as available at the same time. if we are checking ipv4 then should we set IPv6 based on `localCacheStatuses[result.ID].Available.IPv6` and vice versa?"_

unless I'm misunderstanding what he's saying

Reply:
yea the original plan was for each cycle to only update the status of what its checking. so if we are checking IPv6 then the Available.IPv4 should remain as it was and vice versa. it might be as easy as just removing the `availStatus.Available.IPvX &&` from these lines

Reply:
Ahh I misread Matt's original comment. In that case then I think removing the && should fix the issue then.